core/dbt/contracts/graph/nodes.py

```diff
@@ -492,6 +502,40 @@ def depends_on_nodes(self):
     def depends_on_macros(self):
         return self.depends_on.macros
 
+    def build_contract_checksum(self):
+        # We don't need to construct the checksum if the model does not
+        # have contract enabled, because it won't be used.
+        # This needs to be executed after contract config is set
+        if self.contract is True:
+            contract_state = ""
+            # We need to sort the columns so that order doesn't matter
+            # columns is a str: ColumnInfo dictionary
+            sorted_columns = sorted(self.columns.values(), key=lambda col: col.name)
+            for column in sorted_columns:
+                contract_state += f"|{column.name}"
+                contract_state += str(column.data_type)
+            data = contract_state.encode("utf-8")
+            self.contract_checksum = hashlib.new("sha256", data).hexdigest()
+
+    def same_contract(self, old) -> bool:
+        if old.contract is False and self.contract is False:
+            # Not a change
+            return True
+        if old.contract is False and self.contract is True:
+            # A change, but not a breaking change
+            return False
+        breaking_change_reasons = []
+        if old.contract is True and self.contract is False:
+            # Breaking change: throw an error
+            breaking_change_reasons.append("contract has been disabled")
+        if self.contract_checksum != old.contract_checksum:
+            # Breaking change: throw an error
+            breaking_change_reasons.append("column definitions have changed")
+        else:
+            # No change
+            return False
```

Comment:
should this be `True` instead, since there is no change?

Also, it's possible to hit this line and return early even if there are elements in `breaking_change_reasons` from the first case (`contract_checksum` has not changed but `self.contract` changed from `True` to `False`). Perhaps instead of this else statement we need a separate block at the end of this method: 

```python
if breaking_change_reasons: 
   raise (ModelContractError(reasons=" and ".join(breaking_change_reasons), node=self))
else: 
  # No change
  return True
```

Reply:
Yeah, that last bit should be True. But there's another issue here, which is that right now I'm not generating the contract_checksum if contract is false, so we're not actually ever going to get both "contract has been disabled" and "column definitions have changed".  It feels kind of wrong to build the contract_checksum if contract enabled is false, but do you want it anyway?

Reply:
Well, technically it would be a generated_checksum != None. But there might not even be any column changes for that None, so it would be wrong to flag it.

Reply:
I've rearranged it so I think it should work now.